### PR TITLE
qwt: switch to qt6

### PR DESCRIPTION
**Pull request by MehdiChinoune**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/gnuradio.rb [hunk 1]
```diff
@@ -38,7 +38,7 @@ class Gnuradio < Formula
   depends_on "pyqt@5"
   depends_on "python@3.9"
   depends_on "qt@5"
-  depends_on "qwt"
+  depends_on "qwt-qt5"
```
Comment (SMillerDev) [inline Formula/gnuradio.rb:41]:
```suggestion
  depends_on "qwt-qt@5"
```

File: Formula/qwt-qt5.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
+class QwtQt5 < Formula
```
Comment (SMillerDev) [inline Formula/qwt-qt5.rb:1]:
```suggestion
class QwtQtAT5 < Formula
```

File: Formula/qwt-qt5.rb [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  url "https://downloads.sourceforge.net/project/qwt/qwt/6.2.0/qwt-6.2.0.tar.bz2"
+  sha256 "9194f6513955d0fd7300f67158175064460197abab1a92fa127a67a4b0b71530"
+  license "LGPL-2.1-only" => { with: "Qwt-exception-1.0" }
+
+  livecheck do
+    url :stable
+    regex(%r{url=.*?/qwt[._-]v?(\d+(?:\.\d+)+)\.t}i)
+  end
+
+  keg_only "it conflicts with qwt"
```
Comment (SMillerDev) [inline Formula/qwt-qt5.rb:13]:
```suggestion
  keg_only :versioned_formula
```

File: Formula/qwt.rb [hunk 4]
```diff
@@ -10,14 +10,7 @@ class Qwt < Formula
     regex(%r{url=.*?/qwt[._-]v?(\d+(?:\.\d+)+)\.t}i)
   end
 
-  bottle do
-    sha256 cellar: :any, arm64_big_sur: "0538bfe404c21c264efe57fbc36d3cff81c39b86679d67c48501166597ab8cad"
-    sha256 cellar: :any, big_sur:       "14a5fd16a5abcf3a04b3f6d097649fbc1dd51e9fbb50d05f885757a9d9f3d9f9"
-    sha256 cellar: :any, catalina:      "c3a727be657b20efdd6a8ddec980bd28f5367ae41a0a7abefb74af86c1f24e83"
-    sha256 cellar: :any, mojave:        "5bb62a4122ade6485247357b22f0619ff35f518d9ea3f454f05c7d5c4b60985e"
-  end
```
Comment (SMillerDev) [inline Formula/qwt.rb:18]: You shouldn't change the bottle block here.

File: Formula/qwt-qt5.rb [hunk 5]
```diff
@@ -0,0 +1,103 @@
+class QwtQtAT5 < Formula
```
Comment (SMillerDev) [inline Formula/qwt-qt5.rb:1]: The file also needs to be renamed to `qwt-qt@5.rb` for this

Comment (MehdiChinoune) [reply Formula/qwt-qt5.rb:1]: I get an error about missing unversioned qwt-qt.rb

**Comment by carlocab** [comment]
Please split your commits so that each commit modifies only one formula:

> - One formula per commit; one commit per formula.
> - Keep merge commits out of the pull request.

When you do so, please format your commit messages according to Homebrew/core style:

> At Homebrew, we like to put the name of the formula up front like so: `foobar 7.3 (new formula)`.

> The preferred commit message format for simple version updates is `foobar 7.3` and for fixes is `foobar: fix flibble matrix.`.

See the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit) referenced in the PR template for more information.

File: Formula/qwt.rb [hunk 6]
```diff
@@ -17,7 +17,7 @@ class Qwt < Formula
     sha256 cellar: :any, mojave:        "5bb62a4122ade6485247357b22f0619ff35f518d9ea3f454f05c7d5c4b60985e"
   end
 
-  depends_on "qt@5"
+  depends_on "qt"
```
Comment (carlocab) [inline Formula/qwt.rb:20]: We need to bump the revision here since there's been a change in dependencies.

File: Formula/gnuradio.rb (hunk 1 above)
Comment (carlocab) [inline Formula/gnuradio.rb:41]: Can we change this all to `qt` and `pyqt` instead?

File: Formula/qcli.rb [hunk 7]
```diff
@@ -17,7 +17,7 @@ class Qcli < Formula
   depends_on "pkg-config" => :build
   depends_on "ffmpeg"
   depends_on "qt@5"
-  depends_on "qwt"
+  depends_on "qwt-qt5"
```
Comment (carlocab) [inline Formula/qcli.rb:20]: Can we switch this to `qt` and keep `qwt` instead?

**Review by carlocab — COMMENTED** [review]
Does `qwtpolar` not need to be switched to `qt` from `qt@5` as well?

File: Formula/gnuradio.rb (hunk 1 above)
Comment (MehdiChinoune) [reply Formula/gnuradio.rb:41]: qwtpolar is now part of qwt.
see: https://sourceforge.net/p/qwtpolar/news/2021/07/qwtpolar-has-become-part-of-the-qwt-library/

File: Formula/qcli.rb (hunk 7 above)
Comment (MehdiChinoune) [reply Formula/qcli.rb:20]: I don't know!
Should I ?

**Comment by MehdiChinoune** [comment]
> Does `qwtpolar` not need to be switched to `qt` from `qt@5` as well?

qwtpolar is now part of qwt (6.2.0)

File: Formula/gnuradio.rb (hunk 1 above)
Comment (MehdiChinoune) [reply Formula/gnuradio.rb:41]: After taking a look at https://github.com/gnuradio/gnuradio/blob/master/gr-qtgui/CMakeLists.txt. It is clear that it supports onlt qt5 and pyqt5.

**Comment by MehdiChinoune** [comment]
`#{Formula["qt"].opt_include}` reports an incorrect value `I/opt/homebrew/opt/qt/lib/QtGui.framework/Versions/5/Headers`, it should be `I/opt/homebrew/opt/qt/lib/QtGui.framework/Versions/6/Headers`

File: Formula/qwt.rb [hunk 8]
```diff
@@ -62,21 +62,21 @@ def install
       system ENV.cxx, "test.cpp", "-o", "out",
         "-std=c++11",
         "-framework", "qwt", "-framework", "QtCore",
-        "-F#{lib}", "-F#{Formula["qt@5"].opt_lib}",
+        "-F#{lib}", "-F#{Formula["qt"].opt_lib}",
         "-I#{lib}/qwt.framework/Headers",
-        "-I#{Formula["qt@5"].opt_lib}/QtCore.framework/Versions/5/Headers",
-        "-I#{Formula["qt@5"].opt_lib}/QtGui.framework/Versions/5/Headers"
+        "-I#{Formula["qt"].opt_lib}/QtCore.framework/Versions/5/Headers",
+        "-I#{Formula["qt"].opt_lib}/QtGui.framework/Versions/5/Headers"
```
Comment (carlocab) [inline Formula/qwt.rb:68]:
```suggestion
        "-I#{Formula["qt"].opt_lib}/QtCore.framework/Versions/6/Headers",
        "-I#{Formula["qt"].opt_lib}/QtGui.framework/Versions/6/Headers"
```
There's nothing wrong with `Formula["qt"].opt_lib`. You just failed to change the `5`s to `6`s.

**Comment by carlocab** [comment]
> `#{Formula["qt"].opt_include}` reports an incorrect value `I/opt/homebrew/opt/qt/lib/QtGui.framework/Versions/5/Headers`, it should be `I/opt/homebrew/opt/qt/lib/QtGui.framework/Versions/6/Headers`

`Formula["qt"].opt_include` évaluâtes to `$HOMEBREW_PREFIX/opt/qt/include`, so there is nothing incorrect about the value it reports.

**Comment by carlocab** [comment]
> > Does `qwtpolar` not need to be switched to `qt` from `qt@5` as well?
> 
> 
> 
> qwtpolar is now part of qwt (6.2.0)

Then we need to disable it and point users to `qwt`.

File: Formula/qwt.rb (hunk 8 above)
Comment (MehdiChinoune) [reply Formula/qwt.rb:68]: Thanks!

**Comment by MehdiChinoune** [comment]
> > > Does `qwtpolar` not need to be switched to `qt` from `qt@5` as well?
> > 
> > 
> > qwtpolar is now part of qwt (6.2.0)
> 
> Then we need to disable it and point users to `qwt`.

How?

File: Formula/qwt.rb [hunk 9]
```diff
@@ -41,11 +42,10 @@ def install
     on_linux do
       spec = "linux-g++"
     end
-    spec << "-arm64" if Hardware::CPU.arm?
     args << spec
 
-    qt5 = Formula["qt@5"].opt_prefix
-    system "#{qt5}/bin/qmake", *args
+    qt = Formula["qt"].opt_prefix
+    system "#{qt}/bin/qmake", *args
```
Comment (carlocab) [inline Formula/qwt.rb:48]:
```suggestion
    qt = Formula["qt"]
    system "#{qt.opt_prefix}/bin/qmake", *args
```
This way we don't need to keep repeated `Formula["qt"]` below and just use `qt`.

**Comment by carlocab** [comment]
You also need these changes in `gnuradio`:
```diff
diff --git a/Formula/gnuradio.rb b/Formula/gnuradio.rb
index 88121be3e4..812e09aad9 100644
--- a/Formula/gnuradio.rb
+++ b/Formula/gnuradio.rb
@@ -103,8 +103,8 @@ class Gnuradio < Formula
       -DENABLE_DEFAULT=OFF
       -DPYTHON_EXECUTABLE=#{venv_root}/bin/python
       -DPYTHON_VERSION_MAJOR=3
-      -DQWT_LIBRARIES=#{Formula["qwt"].lib}/qwt.framework/qwt
-      -DQWT_INCLUDE_DIRS=#{Formula["qwt"].lib}/qwt.framework/Headers
+      -DQWT_LIBRARIES=#{Formula["qwt-qt5"].lib}/qwt.framework/qwt
+      -DQWT_INCLUDE_DIRS=#{Formula["qwt-qt5"].lib}/qwt.framework/Headers
       -DCMAKE_PREFIX_PATH=#{Formula["qt@5"].opt_lib}
       -DQT_BINARY_DIR=#{Formula["qt@5"].opt_bin}
       -DENABLE_TESTING=OFF
```

**Comment by MehdiChinoune** [comment]
gnuradio couldn't build with qwt 6.2.0
gnuradio/gnuradio#4956

**Comment by scpeters** [comment]
I believe any formulae in 3rd-party taps that depend on `qwt` will be broken when this is merged. To simplify their migration path, I've added an alias for `qwt-qt5` formula in https://github.com/Homebrew/homebrew-core/pull/84275 so that 3rd-party taps can change their dependencies to point to `qwt-qt5` in advance. They may still be broken when that is merged, but I think there's a chance it will go more smoothly.

**Comment by MehdiChinoune** [comment]
- Could this pr merged with broken gnuradio?
- Should I backport the fix to gnuradio?
- Should I wait until new release from gnuradio that supports qwt 6.2.0?

**Comment by carlocab** [comment]
We don't want to merge this with `gnuradio` broken, but we can backport a fix if one is available.

**Comment by MehdiChinoune** [comment]
Is there a way to silent that error about qwt-qt5 patch
```
==> brew audit qwt-qt5 --online --new-formula
==> FAILED
==> Installing 'bundler' gem
qwt-qt5:
  * Formulae should not require patches to build. Patches should be submitted and accepted upstream first.
Error: 1 problem in 1 formula detected
```

qwt-qt5 formula is a duplicate of the old qwt!

**Review by carlocab — APPROVED** [review]
Thanks!

**Pull request closed without merging**